### PR TITLE
Refactor user

### PR DESCRIPTION
**Pull request by sielay**
- This is proof of concept for #337.
- It contain fix for #288
- It partially address #288
- It allows multiple emails and usernames
- It keeps back compatibility

It moves all identities to collection in user object. It keeps password and salt shared.

Tested against Facebook and GitHub, unit tests added. Documentation required.

Current user structure
```
{
     _id : ObjectID
     displayName : "My name",
     salt: "<SALT>",
     password: "<HASHED_PASSWORD>",
     identities : [ {
          provider : UserModel.EMAIL,
          id: "test@test.com"
     }, {
          provider : UserModel.USERNAME,
          id : "Tony Stark"
     }, {
          provider : "github",
          id : "<ID>",
          accessToken : "<ACCESS_TOKEN>",
          refreshToken: "<REFRESH_TOKEN>",
          providerData : Object
     } ]
}
```

Useful methods:
```
user.getIdentity('slack');
OR
user.getIdentity('facebook', '<SECOND_ID>');
```
Previously you needed multiple lines to find if that was `user.providerData` or `user.additionalProviderData[provider]`.

Hope you like it.

### CODE REVIEW

**Comment by ilanbiala** [comment]
I actually think a Map would be better suited, because the array is less informative and requires virtuals...

```
identities: {
  google: {
    id: "test@gmail.com",
    password: "abc123" // plain text just for simplicity
  },
  facebook: {
    id: "7778889999",
    password: "abc123"
  }
}
```
With this one you can just do `user.identities.google.id`. Wouldn't that be more convenient than having to do `user.identities[0].id`? I'm assuming that `google` exists, because in both cases you have to check for it, but still checking for the existence in an object is as simple as `if (user.identities.google)`, it reads cleaner than a for loop.

**Comment by sielay** [comment]
Yes, but that works fine when you assume there is one identity of type per user. What may be not case. Some apps will look for np connecting multiple twitter account. Also how will we keep same structure for emails, allowing having multiple email addresses at the same time? Array base is not perfect, but most flexible. Proven in many previous projects.

**Comment by ilanbiala** [comment]
@sielay I did forget about multiple accounts for the same provider, but there you could probably get around that by using part of the provider data, no? I'm just trying to make it easier and and cleaner.

Any thoughts on possible keys inside the provider object? So using the email/username like `user.identities.google['abc@gmail.com']` would be one possibility. Are there better ones?

**Comment by sielay** [comment]
Got your point. Wonder about JSON compatibility and how mongo will behave. Some implementations complain about special characters in keys. Let's wait for @roieki and others input. I can then align to whatever form will work for you. Can also create nice virtuals to help (current are for keeping back compatibility). As you see I can code fast :) Just want to code think that will work for all ;)

**Comment by roieki** [comment]
Sorry guys - I won't be able to go over this in the coming week - @amoshaviv @NeverOddOrEven @lirantal - could you please take a look?

**Comment by NeverOddOrEven** [comment]
There was a discussion about this in a different now-closed issue: https://github.com/meanjs/mean/pull/290

Resume conversation here please. I will review once I have some availability.

**Comment by sielay** [comment]
Thanks. @NeverOddOrEven . I will do small cleanup and refactor once we all agree on arrayVSobject structure. Before that no detailed checkup has sense.

**Comment by ilanbiala** [comment]
@lirantal I think object is easier to work with if we can figure out a good structure, no?

**Comment by sielay** [comment]
I made some extra fixes to bugs I figured that in production project. I will update PR and test object. I am afraid of special chars in keys, need to see how mongo behaves. Will test and we'll see.

**Comment by ilanbiala** [comment]
@sielay sounds good, can't wait to see the changes.

**Comment by sielay** [comment]
Hi. I checked. Object structure would be in conflict with Mongo https://jira.mongodb.org/browse/SERVER-3229?jql=labels%20%3D%20syntax so doing that brings risks (even if by some chance it would work) @ilanbiala

**Comment by sielay** [comment]
Anyhow I added support for multiple emails to settings page. I leave it at the moment to review, need to go to project which I base on this PR (second, first http://pulzair.com is kind of live ;)). Feel free to discuss.

**Comment by ilanbiala** [comment]
@sielay there have been a couple changes in the user controller and a couple more will probably be merged in before this PR just because of they are smaller and simpler. @lirantal @NeverOddOrEven @roieki @amoshaviv what do you guys think of the current PR?

**Comment by sielay** [comment]
@ilanbiala no problem. I will rebase commit and retest it after those changes, just give me note when. I have to maintain this fork as I have projects basing on such structure.

File: package.json [hunk 1]
```diff
@@ -1,73 +1,75 @@
...
-		"grunt-karma": "~0.9.0",
-		"load-grunt-tasks": "~1.0.0",
-		"karma": "~0.12.0",
-		"karma-jasmine": "~0.2.1",
-		"karma-coverage": "~0.2.0",
-		"karma-chrome-launcher": "~0.1.2",
-		"karma-firefox-launcher": "~0.1.3",
-		"karma-phantomjs-launcher": "~0.1.2"
-	}
+  "name": "meanjs",
```
Comment (ilanbiala) [inline package.json]: Can you revert these whitespace changes?

File: config/strategies/local.js [hunk 2]
```diff
@@ -14,9 +14,7 @@ module.exports = function() {
 			passwordField: 'password'
 		},
 		function(username, password, done) {
-			User.findOne({
-				username: username
-			}, function(err, user) {
+			User.findByProvider([User.USERNAME,User.EMAIL], username, function(err, user) {
```
Comment (ilanbiala) [inline config/strategies/local.js]: Space between comma, why uppercase?

File: app/controllers/users/users.authentication.server.controller.js [hunk 3]
```diff
@@ -97,78 +97,17 @@ exports.oauthCallback = function(strategy) {
...
-				return done(err, user, '/#!/settings/accounts');
-			});
-		} else {
-			return done(new Error('User is already connected using this provider'), user);
-		}
+exports.saveOAuthUserProfile = function(err, user, isNew, done) {
+	if(err) {
+		return done(err);
+	}
+	if(!user) {
```
Comment (ilanbiala) [inline app/controllers/users/users.authentication.server.controller.js]: spacing.

File: app/controllers/users/users.profile.server.controller.js [hunk 4]
```diff
@@ -26,7 +26,18 @@ exports.update = function(req, res) {
 		user.updated = Date.now();
 		user.displayName = user.firstName + ' ' + user.lastName;
 
-		user.save(function(err) {
+        user.identities.forEach(function(auth){
```
Comment (ilanbiala) [inline app/controllers/users/users.profile.server.controller.js]: spacing after ).

File: app/controllers/users/users.profile.server.controller.js [hunk 5]
```diff
@@ -26,7 +26,18 @@ exports.update = function(req, res) {
...
+            auth.accessToken = auth.accessToken || null;
+            auth.refreshToken = auth.refreshToken || null;
+            auth.providerData = auth.providerData || null;
+        });
+
+        if(req.body.username) {
```
Comment (ilanbiala) [inline app/controllers/users/users.profile.server.controller.js]: spacing.

**Comment by ilanbiala** [comment]
@sielay sorry it took so long, just added some comments. You might want to run it through JSHint to properly format everything to be consistent with the rest of MEAN.js.

**Comment by sielay** [comment]
@ilanbiala likely I will share something. Pulzair.com is using that structure so far.

File: config/strategies/local.js (hunk 2 above)
Comment (sielay) [inline config/strategies/local.js]: UPPERCASE are for static `consts`.

**Comment by ilanbiala** [comment]
@sielay does this account for unique email support like in #439?

**Comment by sielay** [comment]
@ilanbiala suppose. Wasn't there test for that already?

**Comment by ilanbiala** [comment]
The unique property isn't set on the email in 0.3 or 0.4, but now we need to check across multiple elements in the array, right?

File: config/strategies/facebook.js [hunk 6]
```diff
@@ -19,23 +20,20 @@ module.exports = function() {
...
-				username: profile.username,
-				provider: 'facebook',
-				providerIdentifierField: 'id',
-				providerData: providerData
+				username: profile.username
 			};
 
 			// Save the user OAuth profile
-			users.saveOAuthUserProfile(req, providerUserProfile, done);
+			User.oAuthHandle(req.user, 'facebook', providerData.id, accessToken, refreshToken, providerData, userData, function(err, user, isNew) {
```
Comment (miguelcoba) [inline config/strategies/facebook.js]: Nice simplification using the oAuthHandle() method

File: app/models/user.server.model.js [hunk 7]
```diff
@@ -122,25 +145,149 @@ UserSchema.methods.authenticate = function(password) {
...
+                var possibleUsername = userData.username || ((userData.email) ? userData.email.split( '@' )[ 0 ] : '');
+                return UserModel.findUniqueUsername( possibleUsername, null, function ( availableUsername ) {
+                    user = new UserModel( userData );
+                    user.setIdentity( provider, id, accessToken, refreshToken, providerData );
+                    user.save( function ( err ) {
+                        return callback( err, user, true );
+                    } );
+                } );
+            }
+            return callback( new Error( 'You shoun\'t go to that line, ever' ), user, false );
```
Comment (miguelcoba) [inline app/models/user.server.model.js]: What does it means if execution reaches this line?

File: public/modules/users/controllers/settings.client.controller.js [hunk 8]
```diff
@@ -18,7 +53,14 @@ angular.module('users').controller('SettingsController', ['$scope', '$http', '$l
 
 		// Check if provider is already in use with current user
 		$scope.isConnectedSocialAccount = function(provider) {
-			return $scope.user.provider === provider || ($scope.user.additionalProvidersData && $scope.user.additionalProvidersData[provider]);
+			var connected = false;
+			angular.forEach($scope.user.identities, function(identity) {
+				if(identity.provider === provider ) {
+					connected = true;
+					$scope.linkedProviders.push(provider);
```
Comment (miguelcoba) [inline public/modules/users/controllers/settings.client.controller.js]: Why this $scope.linkedProviders.push(provider) line here. Seems like a sideeffect when the method only should check if user already has an identity with that provider.
Maybe populating this array on loading (like the $scope.identitiesToEmails() call) is a better approach and here only return the result of a $scope.linkedProviders.contains(provider).

**Comment by miguelcoba** [comment]
In general it seems to me a very comprehensive way to store and manage multiple identities associated with a user. The multiple emails, all with provider 'email' is nice.
One thing that must be done is to change all the formatting from spaces to tabs, as the current code is.
I don't think this will be affecting the current use cases for the authentication module of meanjs.org and enables some more login modes for a bigger range of applications.

File: config/strategies/facebook.js (hunk 6 above)
Comment (ilanbiala) [inline config/strategies/facebook.js]: Too many function arguments though I think...

File: app/models/user.server.model.js [hunk 9]
```diff
@@ -122,25 +145,149 @@ UserSchema.methods.authenticate = function(password) {
...
+    }
+};
+
+
+
+
+UserSchema.statics.USERNAME = 'username';
+UserSchema.statics.EMAIL    = 'email';
+
+UserModel = mongoose.model('User', UserSchema);
```
Comment (ilanbiala) [inline app/models/user.server.model.js]: Please keep it consistent as `User`.

**Comment by lirantal** [comment]
quite some changes here. any chance we can have a PR on the 0.4 branch instead of master? we'll definitely be getting a lot of conflicts when trying to merge 0.4 to master.

**Comment by ilanbiala** [comment]
@lirantal you think just 1 PR on 0.4?

**Comment by ilanbiala** [comment]
@sielay when you create the PR for 0.4, please keep a reference to this one just so we can refer back to the conversation.

**Comment by lirantal** [comment]
Yeah, I think it's "safer" for when we want to merge 0.4 into master.

**Comment by wesleyfsmith** [comment]
I agree with @lirantal, seems like it will be a lot easier to make this change part of migration to 0.4, instead of throwing it in with all of the 0.3 changes currently.

**Comment by ilanbiala** [comment]
@sielay Is it possible for you to recreate this PR for 0.4.0 so that we can merge it into master much easier later on?

**Comment by sielay** [comment]
@ilanbiala I currently have no time to update it, but each time I use my fork I merge it with your changes. Once you release 0.4.0 I will simply update it and retest especially I already found some issues after update.

**Comment by ilanbiala** [comment]
@sielay we want to have the refactored user setup ready for the 0.4.0 release, so is there any way you can create a PR in the next few days with this same code for the 0.4.0 branch?

**Comment by ilanbiala** [comment]
@sielay also if you can squash the commits down to 1 when you create a new PR against 0.4.0, that would be awesome.

**Comment by britztopher** [comment]
@sielay any update on a PR for the 0.4 branch.  I am trying to get auth token ticket finished, but need to know how authorization is going to be reworked in 0.4 branch before I can continue

**Comment by sielay** [comment]
@sielay I am not able to pick it up now. Anyhow current PR works in main. I prefer you go with your stuff and then I re-align PR, or anyone else pick up the change and to it in own flavour.

**Comment by britztopher** [comment]
@sielay I just think that my token based stuff depends a ton on the authorization process.  If authorization changes it will cause a major headache to get working with the change.

**Comment by ilanbiala** [comment]
Yes the token change is pretty heavily based on the user implementation. If @sielay can't make the changes now, maybe @britztopher can fork and we'll help him finish the refactor?

**Comment by britztopher** [comment]
@ilanbiala I can try to fork repo, create feature branch, then cherry pick @sielay's commit for authorization changes and merge it into 0.4.0.  @sielay do you think this merge would be a difficult one.  Also, doing this merge will also let me learn the changes from old auth to new making #389 easier to conceptualize

**Comment by britztopher** [comment]
@sielay how were you testing your PR in master?

**Comment by britztopher** [comment]
@sielay, did you say you were using this code already in your application?  I only ask because I just forked your repo and checked out refactor_user branch and its throwing a bunch of errors.  Am I missing something.  

@ilanbiala I have merged #343 into the 0.4.0 branch, however, things are not working like they should, so I am currently debugging them.

**Comment by sielay** [comment]
@britztopher version in this PR started to fail after some updates in dependencies. I am using version which wasn't applied into this PR. I run tests as they are set up in the project. As I said before, because things have changed a bit in the main project and I had no time to work on it, I think we should at the moment close PR without merging and come back to it afterwards.

**Comment by sielay** [comment]
I will create new PR when I align to 0.4.0

**Comment by sielay** [comment]
@britztopher @ilanbiala 
I created new branch which is merged with 0.4.0. Anyhow because you made very drastic changes and continued to store junk into provideData and additionalProviderData (don't like it very much), I have plenty of conflicts. I am doing now MVP of new project and aligning to that I will be stabilising that. Thanks to that I will be able to test it against actual oauth authorities. Fork will be here and later will be new PR https://github.com/sielay/mean/tree/0.4.0

**Comment by lirantal** [comment]
@sielay any progress on the user refactoring PR for 0.4.0?
you should checkout the latest 0.4.0 branch though for all the updated 'master' merge and other PRs to get the updated version there.

**Comment by sielay** [comment]
@lirantal had no chance to come back to that last few weeks, just came back to the office and not sure when will be able, last refactor/rebase was quite time consuming.

**Pull request closed without merging**